assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties

```diff
@@ -672,3 +672,7 @@ che.infra.kubernetes.async.storage.shutdown_timeout_min=120
 
 # Defines the period with which the Asynchronous Storage Pod stopping ability will be performed (once in 30 minutes by default)
 che.infra.kubernetes.async.storage.shutdown_check_period_min=30
+
+# Bitbucket endpoint used for both factory and OAuth integrations. Should contain bitbucket server
+# URL or NULL if no integration expected.
+bitbucket.server.endpoint=NULL
```

Comment:
could it be multiple endpoints?

Reply:
no

Reply:
What if the user wants to use 2 different bitbucket servers for whatever reason?

Also, if I understand the purpose of this property correctly, it is only there to distinguish bitbucket from github by forcing the two to only match different hosts. 

IMHO that is a rather limiting approach. It needlessly limits the factory to only support single hostname per "repository type". Wouldn't it be better to try to somehow figure out what kind of server it is when determining the validity of the URL by e.g. trying to deduce an endpoint which would reveal what type of server it is?

E.g. github could be detected by trying to access `<API_HOST>/octocat`, bitbucket could be determined by trying out a URL specific to its API and if we get auth error we may assume it is bitbucket, if we get 404, it is not a bitbucket server.

Alternatively (or rather, in addition), we could just let the user tell us what kind of server she points to with the URL sent to the factory.

Reply:
Again, this is not the general purpose feature. it is about to satisfy the concrete customer request. If you have read description, it is clearly says that this implementation ONLY for indiviual onprem installations, so i can hardly imagine someone have 2 or 3 different bitbuckets in their "private" network. IMO no much sense to write an heuristic analyzis of the servers when it will be never used in such way (because it you set CRW for your on-prem BB, you will 99% probably not need another one, and even github too).

Reply:
I would consider the thing that @metlos suggested only as a fallback option, because it might significantly increase the factory resolution time. So it could be something like "we know these factories, this is how we can detect them from url string only. If not found, then try some super-smart autodetection". However, I agree with @mshaposhnik that it is different issue.
Anyway, making this property multiple endpoints should be straight-forward. Just make it comma-separated list of endpoints and separate in the `BitbucketURLParser` constructor. Or is there more complexity?

Reply:
So guys let's imagine how the future OAuth  and secrets configuration for multiple servers will look like (that's what i know at the current state and can predict. Maybe there will be something else):

1) servers URL-s for factory detection:
`bitbucket.servers.endpoints=bitbucket-1.com, bitbucket-2.com`

During factory resolve, we will jus try tot match them one by one.

2) OAuth conf for each server:

Server 1:

```
bitbucket.endpoint=https://bitbucket-1.com
oauth.bitbucket.consumerkey=key1111
oauth.bitbucket.privatekey=MII....<long key>...Fdh=
```

Server 2:

```
bitbucket.endpoint=https://bitbucket-2.com
oauth.bitbucket.consumerkey=key2222
oauth.bitbucket.privatekey=a8da....<long key>...ssd=
``` 

So, of course it will not work that way, and we also cannot put some variable part into property key (like `oauth.bitbucket-1.consumerkey=..`). So in that case, we need to think about complete move this configuration into other place (to the NS secrets ?)
And from code part, we will be not able to bind multiple OAuth authenticators of the same type, they will have to be somehow stored in lists and find corresponding on demand.

3) Storing of personal tokens and Git secrets:
So from what was proposed at research, personal token secrets should be also marked for which server it is, and should have unique names:

```
apiVersion: v1
kind: Secret
metadata:
  name: bitbucket-personal-access-token-secret-bitbucket-1.com
  labels:
    app.kubernetes.io/part-of: che.eclipse.org
    app.kubernetes.io/component: bitbucket-personal-access-token
  annotations:
    che.eclipse.org/expired-after: 12323423
    che.eclipse.org/bitbucket-userid: 34593094509
    che.eclipse.org/bitbucket-username: 'tshevchenko'
>>>  che.eclipse.org/bitbucket-server-url: 'https://bitbucket-1.com'
data:
  token: ZHNhZmFkZmFzZGZxMzRyMzRyMTM0
```

so probably we will need and retrieve all and filter by correct hostname.  The same is applicable for Git credentials secret.

@skabashnyuk @metlos

Reply:
I think that delegating the configuration to Namespace Secrets, with proper labels and annotations to match them, is feasible and elegant solution of this. Still question what to do with public servers, that will not require private Secrets.

Reply:
@sparkoo If you mean OAuth configuration... it won't be easy anyway, that conf should be created in the Che server namespace, and OAuth java components (authenticators etc) have to be able to read this conf at server startup time (as they do that when reading from che.properties).

Reply:
I like the idea to store server ulr https://github.com/eclipse/che/pull/18423#discussion_r529737860 in secret.  We can reuse it later.
I think we can skip the idea of multiple endpoints at this stage.

Reply:
> and OAuth java components (authenticators etc) have to be able to read this conf at server startup time (as they do that when reading from che.properties).

why would they have to do it on startup? Why not on demand (with some ttl'd cache)?

Reply:
I share the @sparkoo 's sentiments about this - IMHO storing the config as secrets/configmaps is both consistent with how we store other info and more flexible for the admins...

Reply:
Maybe i am somewhat old-minded, but for me creating the authenticators and other stuff "on-the fly" (means, actually, when user clicked a factory and waits for redirect to provider auth page) by reading the secrets is little inconvenient. It may be boring long, it may fail, and it cannot be stored in Guice (since injector already initialized). So we will need  to keep them somewhere and manage by ourselves.

Reply:
I'm a bit lost here. I must admit that I don't fully understand how oauth with bitbucket works, but my assumption was that every user has to have personal auth token. If so, we can either store the user's token in the database or in Secret and Secret gives Us AND Users convenient way how configure the authentication for "whatever" 3rd party app.